rfcs/20200525-elasticity-and-fault-tolerance-in-multiworkermirroredstrategy.md

```diff
@@ -0,0 +1,439 @@
+
+The standard pattern for updating variables is to:
+
+1. In your function passed to tf.distribute.Strategy.experimental_run_v2, compute a list of (update, variable) pairs. For example, the update might be the gradient of the loss with respect to the variable.
+2. Switch to cross-replica mode by calling tf.distribute.get_replica_context().merge_call() with the updates and variables as arguments.
+3. Call tf.distribute.StrategyExtended.reduce_to(VariableAggregation.SUM, t, v) (for one variable) or tf.distribute.StrategyExtended.batch_reduce_to (for a list of variables) to sum the updates. and broadcast the result to the variable's devices.
+4. Call tf.distribute.StrategyExtended.update(v) for each variable to update its value.
+
+### [Roles](https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras#multi-worker_configuration)
+
+There are two components of `TF_CONFIG`: `cluster` and `task`. `cluster` provides information about the training cluster, which is a dict consisting of different types of jobs such as `worker`. In multi-worker training, there is usually one `worker` that takes on a little more responsibility like saving checkpoint and writing summary file for TensorBoard in addition to what a regular `worker` does. Such worker is referred to as the 'chief' worker, and it is customary that the `worker` with `index` 0 is appointed as the chief `worker` (in fact this is how `tf.distribute.Strategy` is implemented). `task` on the other hand provides information of the current task. The first component `cluster` is the same for all workers, and the second component `task` is different on each worker and specifies the `type` and `index` of that worker. 
+
+
+## _**Adding Elasticity**_
+
+To add elasticity and fault tolerance to MWMS, the system needs to be fault tolerant, and expandable. For fault tolerance, certain faults need to be identifiable and handled explicitly such as communication errors, user error, node errors. These faults will have certain impacts to the system which can reduce the system size and force the system to recover a previously known state. To expand, the system needs to be able to identify new nodes and modify the system to use the new nodes.
+
+Currently in TF, keras has implemented a fault tolerance mechanism. When a node gets pre-empted and is able to restart, keras will wait for the node to come back online, and [everyone will restart from the last saved checkpoint](https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras#fault_tolerance). 
+
+### **_Design Summary_**
+
+This design aims to augment the Multi-Worker Mirrored Strategy to add elasticity while avoiding start/initialization times. In the event of cluster shrinkage or expansion, we don’t want to restart the process/session, but want session to adapt to the cluster changes, avoiding re-compilation, re-optimization, and other usual first time costs. 
```

Comment:
Membership changing adds overhead, too. I’m not convinced that the performance gain outweighs the extra design complexity.

It would be better to add some preliminary experiment results and/or a prototype of this design here.

Reply:
One concern for complete session tear down and bring up is the first pass. There is an issue which has stated 3mins for startup is in the expected range, which is a huge perf impact. 

That sounds good, ill start working on getting some pre-lim numbers together.

Reply:
It would be great if we could go deeper and analyze which parts of the initialization overhead are essential, and which parts could be delayed or optimized out. The analysis might help even we decide to go with saving/restoring in-memory checkpoints.